### DISCUSSION TITLE
Layered Rasters with custom colors

### DISCUSSION THREAD

**Discussion by markobud** [opengeos/leafmap#673] [Q&A] [answered]
Hi! 
I'm trying to overlay different rasters, each with a different color scale, but I'm unable to make it work (below my current approach). Do you have any advice on how to move forward?

Thanks,
Marko

```
import os
import leafmap.leafmap as leafmap

landsat = 'landsat.tif'

url1 = "https://open.gishub.org/data/raster/landsat7.tif"
satellite = leafmap.download_file(url1, "landsat.tif")

m = leafmap.Map()

for b,name,cmap_name in zip([2,1,3],['red','green','blue'],['Reds','Greens','Blues']):
    m.add_raster(landsat, bands=b, layer_name=f'Landsat_{name}',cmap = cmap_name)

m
```

**Comment by giswqs** [comment, marked as answer]
Thank you for reporting. This bug has been fixed in #674. You can update the package using `leafmap.update_package()` after the PR is merged. 

![Peek 2024-02-07 22-09](https://github.com/opengeos/leafmap/assets/5016453/3268f09d-96ac-4d9f-a816-459b018842ca)

**Reply by markobud** [reply]
Thanks a lot! Your solution works perfectly now. However, it does not seem to process correctly custom colormaps:

```
import os
import leafmap.leafmap as leafmap
from matplotlib.colors import ListedColormap

landsat = 'landsat.tif'

url1 = "https://open.gishub.org/data/raster/landsat7.tif"
satellite = leafmap.download_file(url1, "landsat.tif")

cmap1 = ListedColormap(['red','blue'],name='test_1')
cmap2 = ListedColormap(['blue','green'],name='test_2')
cmap3 = ListedColormap(['green','red'],name='test_3')

m = leafmap.Map()

for b,name,cmap in zip([2,1,3],['test_1','test_2','test_3'],[cmap1,cmap2,cmap3]):
    m.add_raster(landsat, bands=b, layer_name=f'Landsat_{name}',cmap = cmap)

m
```

**Reply by giswqs** [reply]
@markobud  This seems to be an issue with localtileserver or rio-tiler. The following example does not use leafmap. It can't render the image, either. rio-tiler appears to support custom colormaps. See https://cogeotiff.github.io/rio-tiler/colormap/#default-rio-tilers-colormaps. I have not figured out how to use custom colormaps with localtileserver. Please report back if you figure it out. 

cc @banesullivan 

```python
from localtileserver import TileClient, get_leaflet_tile_layer
from matplotlib.colors import ListedColormap
from ipyleaflet import Map

client = TileClient('https://github.com/giswqs/data/raw/main/raster/landsat7.tif')
colormap = ListedColormap(['red','blue'],name='test_1')
t = get_leaflet_tile_layer(client, indexes=1, colormap=colormap)
m = Map(center=client.center(), zoom=client.default_zoom)
m.add(t)
m
```

**Reply by banesullivan** [reply]
Ah yes, this is the one thing that I was unable to handle in localtileserver's transition from large-image to rio-tiler. Its possible to do this, I just need to account for in localtileserver. 

Sorry for the loss of functionality! I didn't think many people were using custom colormaps like this

**Reply by rbuehler** [reply]
Hello, I am running into this same issue too.
(how to use custom colormaps with localtileserver/rasters in leafmap)
Is there any news on this? 
Thank you very much

**Reply by giswqs** [reply]
See https://leafmap.org/notebooks/103_raster_colormap/

**Reply by rbuehler** [reply]
I had seen this solution, but plotting an image is not the same e.g. for changing to other color maps after, or reading out pixel values. I guess workarounds can be done here too, but it would be much cleaner if one could define custom color maps for the raster. 
Btw, I would have expected that registering the color map with [rio-tiler](https://cogeotiff.github.io/rio-tiler/colormap/#custom-colormaps) works, but it didn't for me.
As there is workarounds I understand it is not high priority, but if there is any change on this I would appreciate a note here. Thanks again.

**Reply by giswqs** [reply]
Try this: 

```python
import matplotlib
import leafmap

colormap = matplotlib.colors.LinearSegmentedColormap.from_list(
    'ndvi', [
        '#422112',
        '#724C01',
        '#CEA712',
        '#FFA904',
        '#FDFE00',
        '#E6EC06',
        '#BACF00',
        '#8BB001',
        '#72A002',
        '#5B8D03',
        '#448102',
        '#2C7001',
        '#176100',
    ],
    256,
)
m = leafmap.Map()
url = "https://github.com/giswqs/data/raw/main/raster/landsat7.tif"
m.add_raster(url, indexes=[1], colormap=colormap)
m
```
<img width="797" height="511" alt="image" src="https://github.com/user-attachments/assets/5f8a83e6-196d-4d37-9edc-ab1316354ab5" />

**Comment by markobud** [comment]
Thanks both for your time! @giswqs and @banesullivan!
